### PR TITLE
[WIP] Show constraint sources in dependency solver errors

### PR DESCRIPTION
**Pull request by 9999years**
First shot at #10519. I think this needs to be scrapped.

### CODE REVIEW

File: cabal-testsuite/PackageTests/NewSdist/MultiTarget/all-test-sute.out
Comment (jasagredo) [inline cabal-testsuite/PackageTests/NewSdist/MultiTarget/all-test-sute.out:1]: Maybe we can take this opportunity to rename this file to `all-test-suite.out` instead? the name has a typo

File: cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.out [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+Downloading the latest package list from test-local-repo
+# cabal v2-build
+Configuration is affected by the following files:
+- cabal.project
+Resolving dependencies...
+Error: [Cabal-7107]
+Could not resolve dependencies:
+[__0] next goal: my-lib (user goal)
+[__0] rejecting: my-lib; 2.0, 1.0
+      (constraint from cabal.project requires ==0.9)
```
Comment (jasagredo) [inline cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.out:11]: I have not checked the code, but this output looks very good. I don't know why the newline though instead of printing it in the same line.

**Comment by ulysses4ever** [comment]
@jasagredo just a heads-up: on the last Cabal meeting @9999years asked to not comment on their PRs while a PR is in the draft mode. I remember it because I’m guilty in it more than many!

File: cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.out (hunk 1 above)
Comment (ulysses4ever) [reply cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.out:11]: I quite like the new line fwiw. Of course, it’s impossible to find a consensus around such things…

Comment (9999years) [reply cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.out:11]: This very nice output formatting is thanks to @philderbeast in #9578 😁

Comment (philderbeast) [reply cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.out:11]: > I don't know why the newline though instead of printing it in the same line.

The rejecting message can be quite long and looks alright with a flow layout but showing project and config constraints with their "imported by" parent chain didn't so I went with an indented layout for those, starting on a new line:

 https://github.com/haskell/cabal/blob/c3a9dd7722335900619d64ad19be79ec5b7cf402/cabal-testsuite/PackageTests/ConditionalAndImport/cabal.out#L242-L253

**Comment by 9999years** [comment]
Needs a release note.

**Comment by 9999years** [comment]
Added a release note and tests, addressed comments. Should be ready to go!

File: cabal-install/cabal-install.cabal [hunk 2]
```diff
@@ -243,6 +243,7 @@ library
         zlib       >= 0.5.3    && < 0.8,
         hackage-security >= 0.6.2.0 && < 0.7,
         text       >= 1.2.3    && < 1.3 || >= 2.0 && < 2.2,
+        transformers >= 0.2    && <0.7,
```
Comment (geekosaur) [inline cabal-install/cabal-install.cabal:246]: Anyone know if this will complicate bootstrapping?

**Review by geekosaur — APPROVED** [review]
It's a pity 3.14.1.0 is breathing down our necks; it would be nice to see this go in, seems like it'd be a big UX improvement.

File: cabal-install/cabal-install.cabal (hunk 2 above)
Comment (9999years) [reply cabal-install/cabal-install.cabal:246]: Should be fine, `Cabal-syntax`, `Cabal`, `cabal-install-solver`, `Cabal-hooks`, and `cabal-bootstrap-gen` already depend on `transformers`.

File: cabal-install-solver/src/Distribution/Solver/Types/WithConstraintSource.hs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  ) where
+
+import Distribution.Solver.Compat.Prelude
+
+import Distribution.Solver.Types.ConstraintSource (ConstraintSource (..), showConstraintSource)
+import Distribution.Pretty (Pretty (pretty))
+import Text.PrettyPrint
+
+-- | A package bundled with a `ConstraintSource`.
+data WithConstraintSource pkg =
```
Comment (jasagredo) [inline cabal-install-solver/src/Distribution/Solver/Types/WithConstraintSource.hs:20]: I see this type being used as `WithContraintSource UserConstraint` so I think `pkg` is not actually a `pkg`? i.e. `constraintPackage` is misnamed?

File: cabal-install/src/Distribution/Client/CmdBuild.hs [hunk 4]
```diff
@@ -135,7 +141,7 @@ defaultBuildFlags =
 -- "Distribution.Client.ProjectOrchestration"
 buildAction :: NixStyleFlags BuildFlags -> [String] -> GlobalFlags -> IO ()
 buildAction flags@NixStyleFlags{extraFlags = buildFlags, ..} targetStrings globalFlags =
-  withContextAndSelectors RejectNoTargets Nothing flags targetStrings globalFlags BuildCommand $ \targetCtx ctx targetSelectors -> do
+  withContextAndSelectors RejectNoTargets Nothing flags (map (\target -> WithConstraintSource{constraintPackage = target, constraintConstraint = ConstraintSourceCommandlineFlag}) targetStrings) globalFlags BuildCommand $ \targetCtx ctx targetSelectors -> do
```
Comment (jasagredo) [inline cabal-install/src/Distribution/Client/CmdBuild.hs:144]: Perhaps a local bind might make this line shorter?

File: cabal-install-solver/src/Distribution/Solver/Types/WithConstraintSource.hs (hunk 3 above)
Comment (9999years) [reply cabal-install-solver/src/Distribution/Solver/Types/WithConstraintSource.hs:20]: I renamed them, how do these look?
- `constraintPackage` → `constraintInner`
- `constraintConstraint` → `constraintSource`

File: cabal-testsuite/PackageTests/NewSdist/MultiTarget/all-test-suite.test.hs
Comment (grayjay) [inline cabal-testsuite/PackageTests/NewSdist/MultiTarget/all-test-suite.test.hs:1]: Did you mean to also change the content of this test?

File: cabal-install-solver/src/Distribution/Solver/Types/WithConstraintSource.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (9999years) [reply cabal-install-solver/src/Distribution/Solver/Types/WithConstraintSource.hs:20]: Yeah, it could in the future. I don't have a preference either way.

File: cabal-install-solver/src/Distribution/Solver/Types/SourcePackage.hs [hunk 5]
```diff
@@ -25,7 +27,7 @@ data SourcePackage loc = SourcePackage
   , srcpkgSource        :: loc
   , srcpkgDescrOverride :: PackageDescriptionOverride
   }
-  deriving (Eq, Show, Generic, Typeable)
+  deriving (Eq, Show, Functor, Generic, Typeable)
```
[the remark this answers is not in the document]
Comment (9999years) [reply cabal-install-solver/src/Distribution/Solver/Types/SourcePackage.hs:30]: It's used here:
https://github.com/haskell/cabal/blob/db60ab4104e314a87959fa3295fab6430b04a4a0/cabal-install/tests/IntegrationTests2.hs#L706-L714

File: cabal-install/src/Distribution/Client/CmdBench.hs [hunk 6]
```diff
@@ -115,8 +121,11 @@ benchAction flags@NixStyleFlags{..} targetStrings globalFlags = do
   baseCtx <- establishProjectBaseContext verbosity cliConfig OtherCommand
 
   targetSelectors <-
-    either (reportTargetSelectorProblems verbosity) return
-      =<< readTargetSelectors (localPackages baseCtx) (Just BenchKind) targetStrings
+    either (reportTargetSelectorProblems verbosity . map constraintInner) return
+      =<< readTargetSelectors
+        (localPackages baseCtx)
+        (Just BenchKind)
+        (map (\target -> WithConstraintSource{constraintInner = target, constraintSource = ConstraintSourceCommandlineFlag}) targetStrings)
```
[the remark this answers is not in the document]
Comment (9999years) [reply cabal-install/src/Distribution/Client/CmdBench.hs:128]: I'll change it if you want. I think it's really unclear which is supposed to be which, and I have no clue what Cabal means when it tells me a constraint comes from a 'user target'. What is `ConstraintSourceCommandlineFlag` for, if not command line arguments?

File: cabal-testsuite/PackageTests/BuildTargets/UseLocalPackage/use-local-version-of-package.out [hunk 7]
```diff
@@ -19,7 +19,8 @@ Resolving dependencies...
 Error: [Cabal-7107]
 Could not resolve dependencies:
 [__0] next goal: pkg (user goal)
-[__0] rejecting: pkg-2.0 (constraint from user target requires ==1.0)
+[__0] rejecting: pkg-2.0
+      (constraint from cabal.project requires ==1.0)
```
[the remark this answers is not in the document]
Comment (9999years) [reply cabal-testsuite/PackageTests/BuildTargets/UseLocalPackage/use-local-version-of-package.out:23]: I would like to improve this, but I want to focus on plumbing the `ConstraintSource` values into the dependency solver in this PR. Once that's done, we can start making messages more detailed and useful. (E.g. I also want them to point to line numbers in `cabal.project`.)

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs [hunk 8]
```diff
@@ -849,7 +855,18 @@ exResolve
...
         | otherwise = []
-      targets' = fmap (\p -> NamedPackage (C.mkPackageName p) []) targets
+      targets' =
+        fmap
+          ( \p ->
+              Named
+                ( WithConstraintSource
+                    { constraintInner =
+                        NamedPackage (C.mkPackageName p) []
+                    , constraintSource = ConstraintSourceUnknown
```
[the remark this answers is not in the document]
Comment (9999years) [reply cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs:865]: I think that the `targets` parameter should be replaced with one that contains constraint source information. Right now, the constraint source is unknown because that information is missing. We don't know what this function is getting called with or where the targets come from.

File: cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.project [hunk 9]
```diff
@@ -0,0 +1,7 @@
+packages: .
+
+-- This is `my-lib` 0.9.
+source-repository-package
+  type: git
+  location: https://github.com/9999years/cabal-testsuite-my-lib.git
+  tag: 9a0af0aa81325c71e744def11db06265840ffb5f
```
[the remark this answers is not in the document]
Comment (9999years) [reply cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.project:7]: Sure, I just wanted a smaller repository so the tests would run faster.

[the remark this answers is not in the document]
Comment (9999years) [reply cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.project:7]: Grep for `location: ` in this repository and you'll see we're already cloning tons of non-Haskell org repos in the tests.

File: cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.out (hunk 1 above)
[the remark this answers is not in the document]
Comment (9999years) [reply cabal-testsuite/PackageTests/ConstraintSource/SourceRepositoryPackage/cabal.out:11]: Agreed.

File: cabal-testsuite/PackageTests/NewSdist/MultiTarget/all-test-suite.test.hs
Comment (9999years) [reply cabal-testsuite/PackageTests/NewSdist/MultiTarget/all-test-suite.test.hs:1]: No, @jasagredo asked me to rename it https://github.com/haskell/cabal/pull/10524#discussion_r1835472657

File: cabal-testsuite/PackageTests/ConditionalAndImport/cabal.out [hunk 10]
```diff
@@ -89,7 +89,13 @@ Configuration is affected by the following files:
...
 - same-filename/noncyclical-same-filename-b.config
     imported by: noncyclical-same-filename-b.project
-Up to date
+Resolving dependencies...
+Build profile: -w ghc-<GHCVER> -O1
+In order, the following will be built:
+ - my-0.1 (lib:my) (configuration changed)
+Configuring my-0.1...
+Preprocessing library for my-0.1...
+Building library for my-0.1...
```
[the remark this answers is not in the document]
Comment (9999years) [reply cabal-testsuite/PackageTests/ConditionalAndImport/cabal.out:98]: The package configuration now includes the project file path:

```haskell
ElaboratedConfiguredPackage
    { elabUnitId = UnitId "my-0.1-inplace"
    -- ...
    , elabPkgSourceLocation = WithConstraintSource
        { constraintInner = LocalUnpackedPackage "<ROOT>/."
        , constraintSource = ConstraintSourceProjectConfig
            ( ProjectConfigPath
                ( "noncyclical-same-filename-a.project" :| [] )
            )
        }
    -- ...
    }
```

Therefore, when we rebuild with a different `--project-file`, it changes the configuration.

[the remark this answers is not in the document]
Comment (geekosaur) [reply cabal-testsuite/PackageTests/ConditionalAndImport/cabal.out:98]: Isn't that correct anyway? If I switch from the default development config to `release` or `validate` then it should rebuild eveything.

[the remark this answers is not in the document]
Comment (9999years) [reply cabal-testsuite/PackageTests/ConditionalAndImport/cabal.out]: Fixed in e180d685b33e7e3ebd12350f08c8953af1cf57e3

IMO it should only rebuild if switching the project file causes other parameters like GHC flags or dependencies to change.

[the remark this answers is not in the document]
Comment (grayjay) [reply cabal-testsuite/PackageTests/ConditionalAndImport/cabal.out]: Thanks! That makes sense. I'm fine with either behavior, since this sounds like an edge case. It looks like the test now only reruns the dependency solver.

**Comment by 9999years** [comment]
@grayjay I fixed a semantic conflict with #10546, mind re-reviewing?

**Comment by grayjay** [comment]
@9999years I still have a question about the overall change from my last review: https://github.com/haskell/cabal/pull/10524#pullrequestreview-2457312613

**Comment by 9999years** [comment]
> Why do the build targets need `ConstraintSource`s?

@grayjay I'm looking at this now and the issue is that build targets (as `String`s or `TargetSelector`s) are turned into instances of `PackageSpecifier UnresolvedSourcePackage` basically immediately, and those are used all over and can include things like `source-repository-packages`, so we definitely want constraint sources for those... we can mark them as unknown constraint sources but that doesn't lead to great UX, so it seems like it's easiest to just add `ConstraintSourceCommandlineFlag` or `ConstraintSourceImplicitTarget` to the target arguments as early as possible.

**Comment by grayjay** [comment]
@9999years I realized that I forgot to clarify the type of target that I was talking about in the comment above and at the Cabal meeting. It seems like Cabal uses "target" in multiple ways due to the change from v1 to v2 commands:

1. solver targets: The dependency solver takes a set of package names as an argument and solves for those packages and their dependencies.
2. v1 command build targets: v1 build targets had a relatively simple syntax. Those targets were passed to the dependency solver with very little modification.
3. v2 command build targets: v2 build targets have a syntax that is much more complex and expressive than v1 targets. For example, they can specify all packages, a component type, or a file. If I understand correctly, for most commands, the build targets only specify the subset of the project to build and are not passed to the dependency solver. However, cabal still needs to pass targets to the solver. I think that those targets are derived from the structure of the project, such as local packages or source repository packages listed in `cabal.project`. They usually shouldn't change between build commands.

I would expect the solver targets to be associated with constraints, because cabal needs to constrain the versions to match the versions of local packages, when present. I wouldn't expect the v2 targets to be used for any constraints, since they don't affect solving (except for v2 install).

I can see why `PackageSpecifier`s might need to be associated with `ConstraintSource`s in some cases, but I was wondering why this PR adds `ConstraintSource`s to v2 build targets. Here are two specific examples:

- This change seems to add `ConstraintSource`s to targets in the build command: https://github.com/haskell/cabal/pull/10524/files#diff-1de9e22a937e0b39047b8f4cb7695e4b70a68245ef9a8bc3282114d8856b1225R195-R203
- This PR contains a lot of changes involving the `TargetSelector` type, which represents v2 build targets. This change appears to add `ConstraintSource`s to the parsing of `TargetSelector`s: https://github.com/haskell/cabal/pull/10524/files#diff-ff2f819843c952b293028168b4e8f610ca4f5ea554445d813e6bcb4a17e190f8R258-R268

**Comment by ulysses4ever** [comment]
I'm thinking maybe we should bite the bullet and merge it, and be ready to revert if something goes horribly wrong for the users after the next release. I just really don't like the impasse I'm observing :-) But then again, I have minimal expertise.

**Comment by 9999years** [comment]
Yeah, I wish it was easier to make this change in a less invasive fashion :/

**Comment by grayjay** [comment]
I'm not worried about the correctness of the code, but rather the technical debt. I'm concerned that this PR is very large (+1779, -1009, ignoring tests) and adds a lot of code that appears unnecessary and untested (adding constraint source information to the build targets of v2 commands, other than `install`).

I spent more time reading the code, and my best guess as to why this change touched so many lines is the change to `PackageSpecifier`. That type is used by the code that finds all packages in a project as well as the code that reads v2 build targets. Moving the `ConstraintSource` out of `TargetSelector` and/or related types and only associating a `ConstraintSource` with packages where needed could help reduce the amount of code.

**Comment by Mikolaj** [comment]
@9999years: this is a lot of great work (and with many new tests!), but I wonder, in the interest of increasing development velocity, would you consider the extra effort of splitting this into multiple PRs that make reviewing (and, in the future, debugging and possibly reverting) easier? I guess, one PR per well-defined functional change or a class of code changes? I think, most/all tests can be committed in the PR intended to be merged first marked expected_failure as needed. Would there be any technical problems with such a split?
